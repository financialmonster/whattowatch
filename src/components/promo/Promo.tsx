import React, { FC } from 'react';
import { Map } from 'immutable';

type TPromoProps = {
    promo: boolean | object | null | undefined
}

const Promo: FC<TPromoProps> = ({promo}) => {
    const isPromo = (promo: boolean | object | null | undefined): promo is Map<string, any> => {
        return (promo as Map<string, any>).get !== void 0;
    }

    let previewImage: string | undefined, name: string | undefined, genre: string | undefined,
        released: number | undefined;

    if(isPromo(promo)) {
        previewImage = promo.get(`preview_image`);
        name = promo.get(`name`);
        genre = promo.get(`genre`);
        released = promo.get(`released`);
    }

    return (
        <div className="movie-card__wrap">
            <div className="movie-card__info">
                <div className="movie-card__poster">
                    <img src={previewImage} alt={`${name} poster`} width="218" height="327" />
                </div>
                <div className="movie-card__desc">
                    <h2 className="movie-card__title">{ name }</h2>
                    <p className="movie-card__meta">
                        <span className="movie-card__genre">{ genre }</span>
                        <span className="movie-card__year">{ released }</span>
                    </p>
                    <div className="movie-card__buttons">
                        <button className="btn btn--play movie-card__button" type="button">
                            <svg viewBox="0 0 19 19" width="19" height="19">
                                <use xlinkHref="#play-s"></use>
                            </svg>
                            <span>Play</span>
                        </button>
                        <button className="btn btn--list movie-card__button" type="button">
                            <svg viewBox="0 0 19 20" width="19" height="20">
                                <use xlinkHref="#add"></use>
                            </svg>
                            <span>My list</span>
                        </button>
                    </div>
                </div>
            </div>
        </div>
    );
}

export default React.memo(Promo);
