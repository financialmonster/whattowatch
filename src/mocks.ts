import { Map, fromJS, List } from 'immutable';

export const createStoreMock = (mode: string = `common`) => {
    return {
        promo: Map({
            promo: Map(),
            isPromoFetching: true,
            promoError: null
        }),
        films: Map({
            films: fromJS([{genre: `Thriller`, id: 1}, {genre: `Comedy`, id: 2}]),
            isFilmsFetching: mode === `detailed`,
            filmsError: null,
            filter: `All genres`
        }),
        auth: Map({
            user: null,
            authError: null,
            isAuthFetching: false,
            authStatusError: null,
            isAuthStatusFetching: false
        }),
        reviews: Map({
            reviews: List(),
            isReviewsFetching: true,
            reviewsError: null
        })
    }
}
