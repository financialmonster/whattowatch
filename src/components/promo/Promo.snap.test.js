/* eslint-disable import/first */
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({id: 5})}
));

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { Promo } from './Promo';
import { createStoreMock, mockedFilm } from 'mocks';

it(`Promo: correctly rendered`, () => {
    const mockStore = configureStore();
    const store = mockStore(createStoreMock());

    const tree = renderer.create(
        <Provider store={store}>
            <Promo promo={mockedFilm} />
        </Provider>        
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
