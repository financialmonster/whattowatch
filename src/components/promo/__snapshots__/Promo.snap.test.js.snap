// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Promo: correctly rendered 1`] = `
<div
  className="movie-card__wrap"
>
  <div
    className="movie-card__info"
  >
    <div
      className="movie-card__poster"
    >
      <img
        alt="undefined poster"
        height="327"
        width="218"
      />
    </div>
    <div
      className="movie-card__desc"
    >
      <h2
        className="movie-card__title"
      />
      <p
        className="movie-card__meta"
      >
        <span
          className="movie-card__genre"
        >
          Comedy
        </span>
        <span
          className="movie-card__year"
        />
      </p>
      <div
        className="movie-card__buttons"
      >
        <button
          className="btn btn--play movie-card__button"
          onClick={[Function]}
          title="Play"
          type="button"
        >
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
          >
            <use
              xlinkHref="#play"
            />
          </svg>
          <span>
            Play
          </span>
        </button>
        <button
          className="btn btn--list movie-card__button"
          onClick={[Function]}
          title="My list"
          type="button"
        >
          <svg
            height="20"
            viewBox="0 0 19 20"
            width="19"
          >
            <use
              xlinkHref="#add"
            />
          </svg>
          <span>
            My list
          </span>
        </button>
      </div>
    </div>
    <div
      className="player"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="sc-AxjAm fcmpGc"
      >
        <div
          className="sc-AxirZ fLbCRQ"
          rotate="rotate(0deg)"
        />
        <div
          className="sc-AxirZ iRnakP"
          rotate="rotate(30deg)"
        />
        <div
          className="sc-AxirZ jJDLSA"
          rotate="rotate(60deg)"
        />
        <div
          className="sc-AxirZ cfISCN"
          rotate="rotate(90deg)"
        />
        <div
          className="sc-AxirZ jYaMud"
          rotate="rotate(120deg)"
        />
        <div
          className="sc-AxirZ kCauMN"
          rotate="rotate(150deg)"
        />
        <div
          className="sc-AxirZ bbuhJx"
          rotate="rotate(180deg)"
        />
        <div
          className="sc-AxirZ RfLVX"
          rotate="rotate(210deg)"
        />
        <div
          className="sc-AxirZ iDvjl"
          rotate="rotate(240deg)"
        />
        <div
          className="sc-AxirZ fchBdo"
          rotate="rotate(270deg)"
        />
        <div
          className="sc-AxirZ fGeryM"
          rotate="rotate(300deg)"
        />
        <div
          className="sc-AxirZ eVSKan"
          rotate="rotate(330deg)"
        />
      </div>
      <video
        className="player__video"
        onCanPlayThrough={[Function]}
      />
    </div>
  </div>
</div>
`;
